0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+but less well with those that use the recently more popular UTF-8 encoding, such as Go or Rust.
+Any Dafny code that interfaces with external code will often have to convert between `string` values and
+native representations of strings, and baking in the assumption of UTF-16 imposes a complexity and performance
+penalty for multiple target environments.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+As of Dafny 4.0, the default semantics of the `char` type have been changed to represent Unicode scalar values instead of UTF-16 code points.
+This means that surrogate code points, meaning those in the range `[0xD800, 0xDFFF]`, are no longer allowed. This also means
+code points that require more than 16 bits, meaning those in the range `[0x10000, 0x1FFFF]`, are now directly representable
+as `char` values.
+
+The definition of the `char` type is controlled by a command-line switch called `/unicodeChar`, where `/unicodeChar:0`
+enables the previous semantics and `/unicodeChar:1` enables the new semantics described above.
+The default prior to Dafny 4.0 is `/unicodeChar:0`, but the default as of Dafny 4.0 is `/unicodeChar:1`.
+This change in definition can cause previously verified code to no longer verify,
+since expressions such as `someIntValue as char` or `someCharValue + 'a'` are no longer allowed to result in
+surrogate code points.
+
+This change can also result in behavioral changes that may not be caught by verification,
+as they may not intersect with any explicit specifications.
+Here is an example of a program that will behave differently depending on the value of `/unicodeChar`,
+and hence differently on Dafny 4.0 compared to earlier versions:
+
+```dafny
+method Main() {
+  var s := "Unicode is just so \ud83d\ude0e";
+  print s, "\n"     // "Unicode is just so 😎" (not affected by /unicodeChar)
+  print |s|, "\n"   // /unicodeChar:0 => "21"
+                    // /unicodeChar:1 => "20"
+  print s[19], "\n" // /unicodeChar:0 => "�"
+                    // /unicodeChar:1 => "😎"
+}
+```
+
+A related but largely orthogonal issue has also been addressed.
+Dafny source code files must be encoded in UTF-8,
+but in previous versions of Dafny,
+string and character literals could only contain printable and white-space ASCII characters,
+due to a limitation of the Coco/R parser generator the toolchain uses.
+This has been fixed, and both standard form and verbatim form string literals now allow any Unicode characters.
+A second form of escape sequence accepting a hexadecimal number with up to six digits, `\u{XXXXXX}`, 
+is now provided to support characters outside of the Basic Multilingual Plane
+using their direct Unicode code points instead of using surrogates.
+These changes are fully backwards-compatible and not controlled by the `unicodeChar` flag.
+
+```dafny
+// Several different ways to express the same string literal
+var s1 := "Unicode is just so \ud83d\ude0e";
```

Comment:
I propose we forbid this under `/unicodeChar:1`. After all, neither `\ud83d` nor `\ude0e` falls into the numeric ranges of the new `char`.

Reply:
I didn't see a reason to reject valid string literals using the existing escape pattern, personally. I can make this clearer in the proposal, but we can keep both forms with these semantics:

1. `\uXXXX` - UTF-16 code unit, incorrect use of surrogates rejected by the parser.
2. `\u{X}`...`\u{XXXXXX}` - Unicode scalar value, surrogate range values rejected by the parser.

There is precedent for supporting both, as Go allows both `\uXXXX` and `\UXXXXXXXX`. I liked the variable-length syntax of `\u{X..X}` (also used by Swift) better personally, as exactly requiring exactly eight digits just means the first two will always be `00` :)

I'm open to using `U` for the second form if it helps reduce confusion though.

Reply:
Note we should avoid any implication that a single escape sequence produces a single "character", since neither form can produce 🇨🇦 that way. We will only know that a single `u{X..X}` sequence will produce a single `char` (and that's not ultimately that meaningful anyway).

Reply:
Revision 2 now says we only have `\U{XXXXXX}` with `/unicodeChar:1`

Reply:
Also I realized that in Go `\uXXXX` still specifies Unicode code points rather than UTF-16 code units, that is, it rejects surrogate code points (and hence https://github.com/dafny-lang/dafny/issues/1980 :)